### PR TITLE
Removing explicit ResponseObjects

### PR DESCRIPTION
**Pull request by southeo**
Instead of having separate response objects for Handles, DOIs, DigitalSpecimens, etc, we've moved to a more lightweight JSON response, allowing more code reuse. 

Request objects are still separate for each PID type. There are no plans to change how we ingest requests.

### CODE REVIEW

File: src/main/java/eu/dissco/core/handlemanager/controller/HandleController.java [hunk 1]
```diff
@@ -38,72 +54,124 @@ public class HandleController {
...
+  }
+
+  // Update
+  @PreAuthorize("isAuthenticated()")
+  @PatchMapping(value = "/record")
+  public ResponseEntity<JsonApiWrapper> updateHandleRecord(
+      @RequestBody ObjectNode request)
+      throws InvalidRecordInput, PidResolutionException, IOException, ParserConfigurationException, TransformerException, PidCreationException {
+
+    JsonNode data = request.get("data");
```
[the remark this answers is not in the document]
Comment (southeo) [reply src/main/java/eu/dissco/core/handlemanager/controller/HandleController.java:132]: I'm implementing a function that checks if properties are present.
A wrapper class is a good idea, but since different functions have different requirements - and different combinations of requirements (e.g. "id" needed for resolving/update/archive, "type" for update, "attributes" for creation/archive/update), it would result in a multiple wrapper classes, which seems more trouble than it's worth.

File: src/main/java/eu/dissco/core/handlemanager/repository/StatisticsListener.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
+package eu.dissco.core.handlemanager.repository;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import org.jooq.ExecuteContext;
+import org.jooq.ExecuteType;
+import org.jooq.impl.DefaultExecuteListener;
+
+public class StatisticsListener extends DefaultExecuteListener {
```
[the remark this answers is not in the document]
Comment (southeo) [reply src/main/java/eu/dissco/core/handlemanager/repository/StatisticsListener.java:9]: This was for reporting database statistics with Jooq, it helped with debugging. Not currently in use, though.

File: src/main/java/eu/dissco/core/handlemanager/service/HandleService.java [hunk 3]
```diff
@@ -45,105 +55,540 @@ public class HandleService {
...
+    return pidLink.substring(pidLink.length() - 24);
+  }
+
+  // Batch Creation Json
+
+  public List<JsonApiWrapper> updateRecordBatch(List<JsonNode> requests)
+      throws InvalidRecordInput, PidResolutionException, PidServiceInternalError {
     var recordTimestamp = Instant.now();
-    return handleRep.createDoiRecordBatch(handles, recordTimestamp, handleAttributes);
+    List<byte[]> handles = new ArrayList<>();
```
Comment (samleeflang) [inline src/main/java/eu/dissco/core/handlemanager/service/HandleService.java:150]: Can't these three list be combined into one as they hold values of an object, If you create an object with in it the Handle, Type and attributes

File: src/main/java/eu/dissco/core/handlemanager/service/HandleService.java [hunk 4]
```diff
@@ -45,105 +55,540 @@ public class HandleService {
...
+    handleRep.updateRecordBatch(recordTimestamp, attributesToUpdate);
+    var updatedRecords = fetchResolvedRecords(handles);
+
+    List<JsonApiWrapper> wrapperList = new ArrayList<>();
+    int i = 0;
+
+    for (JsonNode updatedRecord : updatedRecords) {
+      String pidLink = updatedRecord.get(PID).asText();
+      String pidName = getPidName(pidLink);
+      JsonApiData jsonData = new JsonApiData(pidName, recordTypes.get(i++), updatedRecord);
```
Comment (samleeflang) [inline src/main/java/eu/dissco/core/handlemanager/service/HandleService.java:182]: The recordType.get(i++) is strange here, can't you get this from the updatedRecord?

File: src/main/java/eu/dissco/core/handlemanager/service/HandleService.java [hunk 5]
```diff
@@ -45,105 +55,540 @@ public class HandleService {
...
-  public List<DigitalSpecimenResponse> createDigitalSpecimenBatch(
-      List<DigitalSpecimenRequest> requests)
-      throws PidResolutionException, ParserConfigurationException, JsonProcessingException, TransformerException, PidCreationException {
-    List<byte[]> handles = hf.genHandleList(requests.size());
-    List<HandleAttribute> handleAttributes = new ArrayList<>();
 
-    for (int i = 0; i < requests.size(); i++) {
-      handleAttributes.addAll(
-          prepareDigitalSpecimenRecordAttributes(requests.get(i), handles.get(i)));
+  public JsonApiWrapper updateRecord(JsonNode request, byte[] handle, String recordType)
```
Comment (samleeflang) [inline src/main/java/eu/dissco/core/handlemanager/service/HandleService.java:190]: Can't an update of record be seen as a updateRecordBatch of list size 1? It would reduce the amount of code to maintain

File: src/main/java/eu/dissco/core/handlemanager/service/HandleService.java [hunk 6]
```diff
@@ -45,105 +55,540 @@ public class HandleService {
...
+    byte[] handle = hf.genHandleList(1).get(0);
     var recordTimestamp = Instant.now();
-    return handleRep.createDigitalSpecimenBatch(handles, recordTimestamp, handleAttributes);
+    ObjectNode dataNode = (ObjectNode) request.get(NODE_DATA);
+    String type = dataNode.get(NODE_TYPE).asText();
+    List<HandleAttribute> newRecord;
+
+    try {
+      switch (type) {
+        case RECORD_TYPE_HANDLE -> {
```
Comment (samleeflang) [inline src/main/java/eu/dissco/core/handlemanager/service/HandleService.java:223]: Brackets are not necessary, see https://openjdk.org/jeps/361

File: src/main/java/eu/dissco/core/handlemanager/service/HandleService.java [hunk 7]
```diff
@@ -45,105 +55,540 @@ public class HandleService {
...
+    JsonApiData jsonData = new JsonApiData(new String(handle), RECORD_TYPE_HANDLE,
+        postedRecordAttributes);
+    JsonApiLinks links = new JsonApiLinks(postedRecordAttributes.get(PID).asText());
+
+    return new JsonApiWrapper(links, jsonData);
   }
 
-  public DoiRecordResponse createDoiRecord(DoiRecordRequest request)
-      throws PidResolutionException, ParserConfigurationException, JsonProcessingException, TransformerException, PidCreationException {
+  public JsonApiWrapper createDoiRecordJson(DoiRecordRequest request)
```
Comment (samleeflang) [inline src/main/java/eu/dissco/core/handlemanager/service/HandleService.java:346]: My IDE says this method is never called

File: src/main/java/eu/dissco/core/handlemanager/service/HandleService.java [hunk 8]
```diff
@@ -45,105 +55,540 @@ public class HandleService {
...
+    var postedRecordAttributes = getRecord(handle);
+    JsonApiData jsonData = new JsonApiData(new String(handle), RECORD_TYPE_DOI,
+        postedRecordAttributes);
+    JsonApiLinks links = new JsonApiLinks(postedRecordAttributes.get(PID).asText());
+    return new JsonApiWrapper(links, jsonData);
   }
 
-  public DigitalSpecimenResponse createDigitalSpecimen(DigitalSpecimenRequest request)
-      throws PidResolutionException, ParserConfigurationException, JsonProcessingException, TransformerException, PidCreationException {
+  public JsonApiWrapper createDigitalSpecimenJson(DigitalSpecimenRequest request)
```
Comment (samleeflang) [inline src/main/java/eu/dissco/core/handlemanager/service/HandleService.java:360]: My IDE says this method is never called

File: src/main/java/eu/dissco/core/handlemanager/service/HandleService.java [hunk 9]
```diff
@@ -45,105 +55,540 @@ public class HandleService {
...
+    JsonApiData jsonData = new JsonApiData(new String(handle), RECORD_TYPE_DS,
+        postedRecordAttributes);
+    JsonApiLinks links = new JsonApiLinks(postedRecordAttributes.get(PID).asText());
+    return new JsonApiWrapper(links, jsonData);
   }
 
-  public DigitalSpecimenBotanyResponse createDigitalSpecimenBotany(
-      DigitalSpecimenBotanyRequest request)
-      throws PidResolutionException, ParserConfigurationException, JsonProcessingException, TransformerException, PidCreationException {
+  public JsonApiWrapper createDigitalSpecimenBotanyJson(DigitalSpecimenBotanyRequest request)
```
Comment (samleeflang) [inline src/main/java/eu/dissco/core/handlemanager/service/HandleService.java:374]: My IDE says this method is never called

File: src/main/java/eu/dissco/core/handlemanager/service/HandleService.java [hunk 10]
```diff
@@ -45,105 +55,540 @@ public class HandleService {
...
+      String pidLink = updatedRecord.get(PID).asText();
+      String pidName = getPidName(pidLink);
+      JsonApiData jsonData = new JsonApiData(pidName, RECORD_TYPE_TOMBSTONE, updatedRecord);
+      JsonApiLinks links = new JsonApiLinks(pidLink);
+      wrapperList.add(new JsonApiWrapper(links, jsonData));
+    }
+    return wrapperList;
+  }
+
+  public JsonApiWrapper archiveRecord(JsonNode request, byte[] handle)
```
Comment (samleeflang) [inline src/main/java/eu/dissco/core/handlemanager/service/HandleService.java:428]: Same as update/create why not see a single record as a batch of one?

File: src/main/java/eu/dissco/core/handlemanager/service/HandleService.java [hunk 11]
```diff
@@ -45,105 +55,540 @@ public class HandleService {
...
+      throws InvalidRecordInput {
+
+    // Data Ingestion Checks
+    Set<String> requestRecordFields = getRecordFields(requestRecordType);
+    checkFields(keys, requestRecordFields, requestRecordType);
+  }
+
+  private Set<String> getRecordFields(String recordType) throws InvalidRecordInput {
+    switch (recordType) {
+      case RECORD_TYPE_HANDLE -> {
```
Comment (samleeflang) [inline src/main/java/eu/dissco/core/handlemanager/service/HandleService.java:469]: Brackets can be removed

File: src/test/java/eu/dissco/core/handlemanager/controller/HandleControllerTest.java [hunk 12]
```diff
@@ -38,29 +40,27 @@
 import org.springframework.http.ResponseEntity;
 
 @ExtendWith(MockitoExtension.class)
+@Slf4j
```
Comment (samleeflang) [inline src/test/java/eu/dissco/core/handlemanager/controller/HandleControllerTest.java:43]: Normally test classes doesn't have logger

File: src/test/java/eu/dissco/core/handlemanager/controller/HandleControllerTest.java [hunk 13]
```diff
@@ -72,6 +72,76 @@ void testGetAllHandles() throws PidResolutionException {
     assertThat(expectedHandles).isEqualTo(response.getBody());
   }
 
+  @Test
+  void testGetAllHandlesByPidStatus() throws Exception {
+    // Given
+    int pageSize = 10;
+    int pageNum = 1;
+    String pidStatus = "TEST";
+    String handle = HANDLE;
```
Comment (samleeflang) [inline src/test/java/eu/dissco/core/handlemanager/controller/HandleControllerTest.java:81]: My IDE says handle isn't used here. You could also just use the static value if it is necessary

File: src/test/java/eu/dissco/core/handlemanager/controller/HandleControllerTest.java [hunk 14]
```diff
@@ -85,136 +155,167 @@ void testGetAllHandlesFailure() {
...
     // Given
-    List<HandleRecordRequest> request = buildHandleRequestList();
-    List<HandleRecordResponse> responseExpected = buildHandleResponseList();
-    given(service.createHandleRecordBatch(request)).willReturn(responseExpected);
+    List<byte[]> handles = List.of(
+        HANDLE.getBytes(StandardCharsets.UTF_8),
+        HANDLE_ALT.getBytes(StandardCharsets.UTF_8));
+
+    List<JsonNode> requests = new ArrayList<>();
+    for (byte[] handle : handles) {
```
Comment (samleeflang) [inline src/test/java/eu/dissco/core/handlemanager/controller/HandleControllerTest.java:242]: `handle` doesn't seem to be used. Could use `.foreach` here

Comment (samleeflang) [reply src/test/java/eu/dissco/core/handlemanager/controller/HandleControllerTest.java:242]: Same as for below where same structure is used

File: src/test/java/eu/dissco/core/handlemanager/controller/HandleControllerTest.java [hunk 15]
```diff
@@ -230,95 +331,158 @@ void testHello() {
...
-    for (int i = 0; i < REQUEST_LEN; i++) {
-      responseList.add(request);
-    }
-    return responseList;
+  @Test
+  void testUnrecognizedPropertyException() throws Exception {
+    // Given
+    DoiRecordRequest request = genDoiRecordRequestObject();
+    ObjectNode requestNode = genCreateRecordRequest(request, RECORD_TYPE_DOI);
+    String message = "123";
```
Comment (samleeflang) [inline src/test/java/eu/dissco/core/handlemanager/controller/HandleControllerTest.java:462]: Not used

File: src/test/java/eu/dissco/core/handlemanager/repository/HandleRepositoryIT.java [hunk 16]
```diff
@@ -2,38 +2,59 @@
...
 class HandleRepositoryIT extends BaseRepositoryIT {
 
+
   private HandleRepository handleRep;
+  private ObjectMapper mapper;
 
   @BeforeEach
   void setup() {
-    handleRep = new HandleRepository(context);
+    mapper = new ObjectMapper().findAndRegisterModules()
```
Comment (samleeflang) [inline src/test/java/eu/dissco/core/handlemanager/repository/HandleRepositoryIT.java:55]: Could just use `var mapper` instead of the class variable

File: src/test/java/eu/dissco/core/handlemanager/repository/HandleRepositoryIT.java [hunk 17]
```diff
@@ -42,220 +63,373 @@ void destroy() {
...
     // Then
-    var postedRecord = context.selectFrom(HANDLES).fetch();
-    assertThat(responseExpected).isEqualTo(responseReceived);
-    assertThat(postedRecord).hasSize(attributes.size());
+    assertThat(responseReceived).isEqualTo(responseExpected);
   }
 
   @Test
-  void testCreateDigitalSpecimenBatch() throws PidCreationException {
+  void testResolveBatchRecord() throws Exception {
```
Comment (samleeflang) [inline src/test/java/eu/dissco/core/handlemanager/repository/HandleRepositoryIT.java:173]: No Exception is throw here, can be removed

Comment (samleeflang) [reply src/test/java/eu/dissco/core/handlemanager/repository/HandleRepositoryIT.java:173]: Same for a couple of others, IDE should grey these out

File: src/test/java/eu/dissco/core/handlemanager/service/HandleServiceTest.java [hunk 18]
```diff
@@ -54,12 +44,14 @@ class HandleServiceTest {
   @Mock
   private HandleRepository handleRep;
 
-  @Mock
+  @Mock(lenient = true)
```
Comment (samleeflang) [inline src/test/java/eu/dissco/core/handlemanager/service/HandleServiceTest.java:47]: Why is this necessary?

File: src/test/java/eu/dissco/core/handlemanager/service/HandleServiceTest.java [hunk 19]
```diff
@@ -54,12 +44,14 @@ class HandleServiceTest {
...
   private PidTypeService pidTypeService;
 
   @Mock
   private HandleGeneratorService hgService;
 
+  private ObjectMapper mapper;
+
   private Instant instant;
```
Comment (samleeflang) [inline src/test/java/eu/dissco/core/handlemanager/service/HandleServiceTest.java:55]: Can be a lovla variable

File: src/test/java/eu/dissco/core/handlemanager/service/HandleServiceTest.java [hunk 20]
```diff
@@ -89,264 +82,725 @@ void destroy() {
...
+
+
+  @Test
+  void testArchiveRecord() throws Exception {
+    // Given
+    byte[] handle = HANDLE.getBytes(StandardCharsets.UTF_8);
+    JsonNode archiveRequest = genTombstoneRequest();
+    List<HandleAttribute> tombstoneAttributesFull = genTombstoneRecordFullAttributes(handle);
+
+    var databaseResponse = genObjectNodeAttributeRecord(tombstoneAttributesFull);
```
Comment (samleeflang) [inline src/test/java/eu/dissco/core/handlemanager/service/HandleServiceTest.java:485]: Unused variable

File: src/test/java/eu/dissco/core/handlemanager/service/HandleServiceTest.java [hunk 21]
```diff
@@ -89,264 +82,725 @@ void destroy() {
...
+
+    // Then
+    assertThat(responseExpectedFirst).isEqualTo(handles);
+    assertThat(responseExpectedSecond).isEqualTo(handles);
+  }
+
+  // Exceptions
+
+  /*
+    @Test
```
Comment (samleeflang) [inline src/test/java/eu/dissco/core/handlemanager/service/HandleServiceTest.java:544]: Remove commented code

File: src/test/java/eu/dissco/core/handlemanager/service/HandleServiceTest.java [hunk 22]
```diff
@@ -89,264 +82,725 @@ void destroy() {
...
+    });
   }
 
-  private List<DoiRecordResponse> generateBatchDoiResponse() {
-    List<DoiRecordResponse> responseList = new ArrayList<>();
-    for (byte[] h : handlesList) {
-      responseList.add(generateTestDoiResponse(h));
+*/
+  @Test
+  void testInvalidInputExceptionCreateHandleRecord() throws Exception {
```
Comment (samleeflang) [inline src/test/java/eu/dissco/core/handlemanager/service/HandleServiceTest.java:559]: Exception never thrown, can be removed

File: src/test/java/eu/dissco/core/handlemanager/service/HandleServiceTest.java [hunk 23]
```diff
@@ -89,264 +82,725 @@ void destroy() {
...
+    // Then
+    assertThat(exception.getMessage()).contains(PRESERVED_OR_LIVING);
+  }
+
+
+
+  /*
+
+    @Test
+  void testUnrecognizedPropertyException(){
```
Comment (samleeflang) [inline src/test/java/eu/dissco/core/handlemanager/service/HandleServiceTest.java:787]: Remove commented code

File: src/main/java/eu/dissco/core/handlemanager/controller/HandleController.java [hunk 24]
```diff
@@ -38,72 +42,118 @@ public class HandleController {
...
+    return ResponseEntity.status(HttpStatus.OK).body(node);
   }
 
-  @PreAuthorize("isAuthenticated()")
-  @PostMapping(value = "/createRecordBatch", params = "pidType=doi")
-  public ResponseEntity<List<DoiRecordResponse>> createDoiRecordBatch(
-      @RequestBody List<DoiRecordRequest> request)
-      throws PidResolutionException, JsonProcessingException, ParserConfigurationException, PidCreationException, TransformerException {
-    return ResponseEntity.status(HttpStatus.CREATED).body(service.createDoiRecordBatch(request));
+  @GetMapping("/records")
```
[the remark this answers is not in the document]
Comment (southeo) [reply src/main/java/eu/dissco/core/handlemanager/controller/HandleController.java:60]: Hmm this is tricky. I'll make it a POST for now. 

Is there is a definite use case for batch resolution of handles? The cnri handle api doesn't actually have this function. Maybe I can cut this.

File: src/main/java/eu/dissco/core/handlemanager/controller/HandleController.java [hunk 25]
```diff
@@ -38,72 +42,118 @@ public class HandleController {
...
+    return ResponseEntity.status(HttpStatus.OK).body(service.updateRecordBatch(requests));
   }
 
   @PreAuthorize("isAuthenticated()")
-  @PostMapping(value = "/createRecord", params = "pidType=digitalSpecimen")
-  public ResponseEntity<DigitalSpecimenResponse> createDigitalSpecimen(
-      @RequestBody DigitalSpecimenRequest hdl)
-      throws PidResolutionException, JsonProcessingException, ParserConfigurationException, TransformerException, PidCreationException {
-    return ResponseEntity.status(HttpStatus.CREATED).body(service.createDigitalSpecimen(hdl));
+  @DeleteMapping(value = "/record")
```
[the remark this answers is not in the document]
Comment (southeo) [reply src/main/java/eu/dissco/core/handlemanager/controller/HandleController.java:131]: I can do this, but I don't really like how delete (and patch/resolve) are received differently between batch and single requests. Do you think this is a concern?

[the remark this answers is not in the document]
Comment (southeo) [reply src/main/java/eu/dissco/core/handlemanager/controller/HandleController.java:131]: Actually, the request here is mandatory for single requests too. We need the tombstone text as part of provenance. 
PUT requests are meant to replace all existing information in a record, maybe that would be more appropriate? (Technically we retain some information in a tombstone record, e.g. PidIssuer, but it's more accurate than deleting).

File: src/main/java/eu/dissco/core/handlemanager/controller/HandleController.java [hunk 26]
```diff
@@ -125,22 +175,52 @@ public ResponseEntity<List<String>> getAllHandles(
...
-    List<String> handleList = service.getHandlesPaged(pidStatus, pageNum, pageSize);
+    List<String> handleList = service.getHandlesPaged(pageNum, pageSize, pidStatus);
     if (handleList.isEmpty()) {
       throw new PidResolutionException("Unable to resolve pids");
     }
     return ResponseEntity.ok(handleList);
   }
 
-  //Error Handling
+  private void checkRequestNodesPresent(JsonNode requestRoot,
```
[the remark this answers is not in the document]
Comment (southeo) [reply src/main/java/eu/dissco/core/handlemanager/controller/HandleController.java:190]: I agree it's not terribly pretty. I'll add JsonSchemaValidation to the backlog for now

File: src/main/java/eu/dissco/core/handlemanager/controller/HandleController.java [hunk 27]
```diff
@@ -125,22 +186,33 @@ public ResponseEntity<List<String>> getAllHandles(
...
+  @ExceptionHandler(PidServiceInternalError.class)
+  private ResponseEntity<String> pidServiceInternalError(PidServiceInternalError e) {
+    String message;
+    if (e.getExceptionCause() != null) {
+      message = e.getMessage() + ". Cause: " + e.getExceptionCause().toString() + "\n "
+          + e.getExceptionCause().getLocalizedMessage();
+    } else {
+      message = e.getMessage();
+    }
+    return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(message);
```
[the remark this answers is not in the document]
Comment (southeo) [reply src/main/java/eu/dissco/core/handlemanager/controller/HandleController.java:219]: So this exception is thrown when of of these are thrown: IOException, ParserConfigurationException, TransformerException. (The exception acts as a wrapper so I can handle them all together like this.) As I understand it, all 3 of these exceptions are server-side errors, not really anything to do with the user, so I don't think the 400s are suitable. 
I could remove this exception handler and ignore these exceptions?

File: src/main/java/eu/dissco/core/handlemanager/domain/jsonapi/JsonApiWrapper.java [hunk 28]
```diff
@@ -0,0 +1,5 @@
+package eu.dissco.core.handlemanager.domain.jsonapi;
+
+public record JsonApiWrapper(JsonApiLinks links, JsonApiData data) {
```
[the remark this answers is not in the document]
Comment (southeo) [reply src/main/java/eu/dissco/core/handlemanager/domain/jsonapi/JsonApiWrapper.java:3]: https://jsonapi.org/format/#document-top-level 
This says data and links are two different top-level objects, where are you seeing this?

[the remark this answers is not in the document]
Comment (southeo) [reply src/main/java/eu/dissco/core/handlemanager/domain/jsonapi/JsonApiWrapper.java:3]: Are you suggesting there should be two links objects, one top-level for the request and one for each returned record? something like: 


{
    "links": {
           "self":"sandbox.dissco.tech/api/v1/records/20.5000.1025/ABCD"
     }
    "data": {
         "id": "20.5000.1025/ABCD",
          "links": {
                  "self": "hdl.handle.net/20.5000.1025/ABCD"
            }
     }
}

File: src/main/java/eu/dissco/core/handlemanager/controller/HandleController.java [hunk 29]
```diff
@@ -38,72 +51,121 @@ public class HandleController {
...
   }
 
   @PreAuthorize("isAuthenticated()")
-  @PostMapping(value = "/createRecord", params = "pidType=doi")
-  public ResponseEntity<DoiRecordResponse> createDoiRecord(
-      @RequestBody DoiRecordRequest hdl)
-      throws PidResolutionException, JsonProcessingException, ParserConfigurationException, TransformerException, PidCreationException {
-    return ResponseEntity.status(HttpStatus.CREATED).body(service.createDoiRecord(hdl));
+  @PatchMapping(value = "/records")
+  public ResponseEntity<List<JsonApiWrapper>> updateRecords(
```
[the remark this answers is not in the document]
Comment (southeo) [reply src/main/java/eu/dissco/core/handlemanager/controller/HandleController.java:119]: I like having the exceptions on a new line...but I'll change it

File: src/main/java/eu/dissco/core/handlemanager/domain/requests/DigitalSpecimenBotanyRequest.java [hunk 30]
```diff
@@ -2,14 +2,17 @@
...
 import lombok.Getter;
+import lombok.NonNull;
 import lombok.ToString;
 
 @Getter
 @ToString
 @EqualsAndHashCode(callSuper = true)
 public class DigitalSpecimenBotanyRequest extends DigitalSpecimenRequest {
 
+  @NonNull
```
[the remark this answers is not in the document]
Comment (southeo) [reply src/main/java/eu/dissco/core/handlemanager/domain/requests/DigitalSpecimenBotanyRequest.java:13]: It's for error handling and better control if an attribute is missing. Instead of throwing a general NPE (and being unable to find the missing attribute), I can parse which attribute is missing from the request object.

File: src/main/java/eu/dissco/core/handlemanager/repository/StatisticsListener.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (southeo) [reply src/main/java/eu/dissco/core/handlemanager/repository/StatisticsListener.java]: Removed

File: src/main/java/eu/dissco/core/handlemanager/service/HandleService.java (hunk 4 above)
Comment (southeo) [reply src/main/java/eu/dissco/core/handlemanager/service/HandleService.java:182]: Technically yes, but this is faster. updatedRecord is a formatted handle resolution, so its type information is in the following format: 
```
 "digitalObjectType": {
                "pid": "https://hdl.handle.net/20.5000.1025/digital-specimen",
                "pidType": "handle",
                "primaryNamefromPid": "digital-specimen"
            },
 "digitalObjectSubtype": {
                "pid": "https://hdl.handle.net/20.5000.1025/botany-specimen",
                "pidType": "handle",
                "primaryNamefromPid": "botany-specimen"
            },
```
So I'd have to read 2 JsonNode objects to determine type. 
It's also unclear how Handles and DOIs are supposed to be formatted. i.e. is the digitalObjectSubtype null or is the field not present at all? I'm hoping this will be resolved in the next iteration of the data model.

File: src/main/java/eu/dissco/core/handlemanager/controller/HandleController.java (hunk 24 above)
[the remark this answers is not in the document]
Comment (samleeflang) [reply src/main/java/eu/dissco/core/handlemanager/controller/HandleController.java]: I think we introduced this one because we felt the CNRI handle server was missing this.

File: src/main/java/eu/dissco/core/handlemanager/controller/HandleController.java (hunk 25 above)
[the remark this answers is not in the document]
Comment (samleeflang) [reply src/main/java/eu/dissco/core/handlemanager/controller/HandleController.java]: Yes that might be correct. Let's make it PUT

File: src/main/java/eu/dissco/core/handlemanager/domain/jsonapi/JsonApiWrapper.java (hunk 28 above)
[the remark this answers is not in the document]
Comment (samleeflang) [reply src/main/java/eu/dissco/core/handlemanager/domain/jsonapi/JsonApiWrapper.java]: Sorry no I misread. Keep it as a top-level

File: src/main/java/eu/dissco/core/handlemanager/service/HandleService.java (hunk 4 above)
Comment (samleeflang) [reply src/main/java/eu/dissco/core/handlemanager/service/HandleService.java]: Ok let's park this for the next iteration. Still don't like the i++ though, maybe give it a better name

File: src/main/java/eu/dissco/core/handlemanager/service/HandleService.java (hunk 6 above)
Comment (samleeflang) [reply src/main/java/eu/dissco/core/handlemanager/service/HandleService.java:192]: Think SonarLint got this wrong as I struggled with this myself. Think they are needed

File: src/main/java/eu/dissco/core/handlemanager/service/HandleService.java (hunk 3 above)
Comment (southeo) [reply src/main/java/eu/dissco/core/handlemanager/service/HandleService.java]: I can wrap Handle and Type together, that makes sense. 
Attributes are a bit trickier: they're batch-posted to the database, which means theyre processed in a "flat list". Each attribute (i.e. each row in the db) is not associated with a particular handle at time of processing.

File: src/main/java/eu/dissco/core/handlemanager/service/HandleService.java (hunk 4 above)
Comment (southeo) [reply src/main/java/eu/dissco/core/handlemanager/service/HandleService.java]: fixed this, used a <Handle, RecordType> hashmap to get the record type instead of iterating through a list. Above problem still stands, however. "RecordType" comes from the user's request and not information from the database.

File: src/main/java/eu/dissco/core/handlemanager/service/HandleService.java [hunk 31]
```diff
@@ -45,105 +53,481 @@ public class HandleService {
...
+      attributesToUpdate.add(prepareUpdateAttributes(handle, requestAttributes));
+    }
+    checkInternalDuplicates(handles);
+    checkHandlesWritable(handles);
+
+    handleRep.updateRecordBatch(recordTimestamp, attributesToUpdate);
+    var updatedRecords = handleRep.resolveBatchRecord(handles);
+
+    List<JsonApiWrapper> wrapperList = new ArrayList<>();
+    int i = 0;
```
[the remark this answers is not in the document]
Comment (southeo) [reply src/main/java/eu/dissco/core/handlemanager/service/HandleService.java]: removed

File: pom.xml [hunk 32]
```diff
@@ -151,6 +147,10 @@
       <artifactId>junit-jupiter</artifactId>
       <scope>test</scope>
     </dependency>
+    <dependency>
```
[the remark this answers is not in the document]
Comment (southeo) [reply pom.xml:151]: this is necessary with testcontainers; ive updated the scope to "test"

**Review by samleeflang — APPROVED** [review]

**Comment by samleeflang** [comment]
Just put that Sonar issue on False Positive or something. Don't see much of an issue on these Record classes with some more parameters for the constructor
